hazelcast/src/main/java/com/hazelcast/spi/memory/QuickSorter.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.memory;
+
+/** <p>
+ * Base class for QuickSort implementations. Works with an abstract notion of a mutable array of
+ * slots addressed by their index. Makes no assumptions on the size or structure of the slots, or on
+ * the way they are compared.
+ * </p><p>
+ * <strong>Not thread-safe</strong> because the contract for implementing subclasses is stateful:
+ * {@link #loadPivot(long)} must update the internal state which is later accessed by
+ * {@link #isGreaterThanPivot(long)} and {@link #isLessThanPivot(long)}.
+ * </p>
+ */
+public abstract class QuickSorter {
```

Comment:
The idea is to make your own subclass that uses the comparator or anything else. The protected API is oriented towards being simple for the JIT compiler to optimize.